File: go/0128_longest_consecutive_sequence.go
package main

// time: log(n)
// space: log(n)

func longestConsecutive(nums []int) int {
	numSet := make(map[int]bool)
	for _, num := range nums {
		numSet[num] = true
	}

	longest := 0
	for num := range numSet {
		if !numSet[num-1] {
			length := 1
			for numSet[num+length] {
				length++
			}
			if length > longest {
				longest = length
			}
		}
	}
	return longest
}

File: csharp/0128-longest-consecutive-sequence.cs
public partial class Solution {
    public int LongestConsecutive(int[] nums) {
        var set = new HashSet<int>(nums);
        var longest = 0;

        foreach (var n in nums) {
            if (!set.Contains(n - 1)) {
                var sequence = 1;
                while (set.Contains(n + sequence)) {
                    sequence++;
                }

                longest = Math.Max(longest, sequence);
            }
        }

        return longest;
    }
}
